### PR TITLE
Added new feature: calculate z moving distance using Gaussian function

### PR DESCRIPTION
**Pull request by qiaodan97**
for benchmarking: deleted (0, 500), (0, 1000) from scaler list
for new feature: below is the logic and math
IMG_0038
IMG_0040

### CODE REVIEW

File: neo_force_scheme/neo_force_scheme.py [hunk 1]
```diff
@@ -302,10 +319,20 @@ def fit_transform(self,
...
+            or y axis in 2d version of the ploted figure.
+        :param drop_columns_from_dataset: indicate the columns that you do not want to be
+            used for calculating distance matrix. Default is None.
+        :param scaler: the range to which all columns of the dataset should be scaled to.
+            Default is (0, 1). If the user does not want the dataset to be scaled, the input
+            should be None.
+        :param z_axis_moving_range: the range to which points are allowed to move on z axis.
+            Default is (0, 0). This parameter will only be used when fix_column_to_z_projection_axis
+            is not None.
+        :param confidence_interval: indicate the parameter which will be used to calculating the actual
```
Comment (christinoleo) [inline neo_force_scheme/neo_force_scheme.py]: typo: "to calculating" should be "to calculate"

File: neo_force_scheme/neo_force_scheme.py [hunk 2]
```diff
@@ -164,10 +164,20 @@ def _fit(self, X,
...
-                   original_z_axis=None, z_axis_moving_range: Optional[Tuple[float, float]] = (0, 0)):
+                   original_z_axis=None, z_axis_moving_range: Optional[Tuple[float, float]] = (0, 0),
+                   confidence_interval: Optional[float] = 1):
         # iterate until max_it or if the error does not change more than the tolerance
         error = math.inf
 
+        z_score = 1
+        if confidence_interval < 1.0:
+            range_strict_limitation = False
+            get_z_score(confidence_interval)
```
Comment (christinoleo) [inline neo_force_scheme/neo_force_scheme.py]: is this z_score ever used? I see its not attached to any variable

File: neo_force_scheme/neo_force_scheme.py [hunk 3]
```diff
@@ -164,10 +164,20 @@ def _fit(self, X,
 
 
     def _transform(self, X, *, index, total, inplace, n_dimension: Optional[int] = 2, force_projection_dimensions=None,
-                   original_z_axis=None, z_axis_moving_range: Optional[Tuple[float, float]] = (0, 0)):
+                   original_z_axis=None, z_axis_moving_range: Optional[Tuple[float, float]] = (0, 0),
+                   confidence_interval: Optional[float] = 1):
         # iterate until max_it or if the error does not change more than the tolerance
         error = math.inf
 
+        z_score = 1
```
Comment (christinoleo) [inline neo_force_scheme/neo_force_scheme.py]: to what end do we use the z_score?

File: neo_force_scheme/neo_force_scheme.py [hunk 4]
```diff
@@ -164,10 +164,20 @@ def _fit(self, X,
...
+        if confidence_interval < 1.0:
```
Comment (christinoleo) [inline neo_force_scheme/neo_force_scheme.py:173]: is this the only place confidence_interval is used? is it just an if to select whether its the strict bins or the gaussian?

File: neo_force_scheme/neo_force_scheme.py [hunk 5]
```diff
@@ -198,6 +212,7 @@ def transform(
             n_dimension: Optional[int] = 2,
             fix_column_to_z_projection_axis=None,
             z_axis_moving_range: Optional[Tuple[float, float]] = (0, 0),
+            confidence_interval: Optional[float] = 0.99,
```
Comment (christinoleo) [inline neo_force_scheme/neo_force_scheme.py]: I noticed that sometimes the default confidence_interval is 0.99 and sometimes (elsewhere in the code) its 1.0. Is there a reason for that?

File: examples/example_plotly_fixed_axis_moving_in_range.py [hunk 6]
```diff
@@ -9,13 +9,14 @@
...
-                     decay=0.95, max_it=100, cuda=False)
-starting_projection_mode = ProjectionMode.RANDOM
+nfs = NeoForceScheme(metric="euclidean", verbose=True, learning_rate0=0.2,
+                     decay=0.95, max_it=300, cuda=False)
+starting_projection_mode = ProjectionMode.TSNE
 fix_column_to_z_projection_axis = -1
 drop_columns_from_dataset = [-1]
 scaler = (0, 1)
 z_axis_moving_range = (-0.2, 0.2)
+confidence_interval = 0.9
```
Comment (christinoleo) [inline examples/example_plotly_fixed_axis_moving_in_range.py:19]: I imagine there is no difference between confidence_interval=0.9 and 0.99, correct? Just wondering..

File: examples/fixed_axis_tsne.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    ..math::
+        D(x, y) = \sqrt{\sum_i (x_i - y_i)^2}
+    """
+    result = 0.0
+    for i in range(x.shape[0]):
+        result += (x[i] - y[i]) ** 2
+    return np.sqrt(result)
+
+
+def kruskal_stress(distance_matrix, projection):
```
Comment (christinoleo) [inline examples/fixed_axis_tsne.py:247]: could we use the kruskal stress already implemented in the cpu_file? If you copy-pasted this from somewhere, you'll need to also include the licence from the source

File: examples/fixed_axis_tsne.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+    k = 0
+    for i in range(size):
+        for j in range(i, size):
+            distance_matrix[k] = euclidean(data[i], data[j])
+            k = k + 1
+
+    return distance_matrix
+
+
+def euclidean(x, y):
```
Comment (christinoleo) [inline examples/fixed_axis_tsne.py:235]: could we use the euclidean already implemented in the distances file? If you copy-pasted this from somewhere, you'll need to also include the licence from the source

File: examples/fixed_axis_tsne.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+        and the starting projection is randomized using starting_projection_mode and random_state.
+        If specified, this must match n_samples.
+    """
+
+    X = data
+
+    if drop_columns_from_dataset is not None:
+        X = np.delete(data, drop_columns_from_dataset, axis=1)
+
+    return X
```
Comment (christinoleo) [inline examples/fixed_axis_tsne.py:300]: these are the same as elsewhere in the code, right? I imagine we could use them instead of having duplicated code

File: neo_force_scheme/neo_force_scheme.py (hunk 2 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py]: Sorry for the confusion, it should be z_score =  get_z_score(confidence_interval), and then be passed to iteration()

File: neo_force_scheme/neo_force_scheme.py (hunk 3 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py]: z_score is used in the iteration() and is to determine the shape of the gaussian function (), and the details of the logic is mentioned is the pictures above

File: neo_force_scheme/neo_force_scheme.py (hunk 4 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py:173]: Yes, this is the only place that condifence_interval is used, it is meant to be used to both select the mode and calculate the z_score (which determine the shape of gaussian function)

File: neo_force_scheme/neo_force_scheme.py (hunk 5 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py]: The default should be 1.0, while 0,99 should only be the default of get_z_score() because we will never use it if confidence_interval is 1.0. Here in line 215 is a typo. sorry for the inconvenience.

File: examples/example_plotly_fixed_axis_moving_in_range.py (hunk 6 above)
Comment (qiaodan97) [reply examples/example_plotly_fixed_axis_moving_in_range.py:19]: since the z scores (and the shapes of the gaussion function) for 0.9 and 0.99 are different, I think they should be different. for example, if we set the moving range to be (-1, 1), 0.9 means 90% of the area under the line of the gaussian function is between -1 and 1, while 0.99 means 99% of the area under the line is between -1 and 1. More specifically, 0.99 means the peak of the function figure (i.e x = 0) should be more sharp, if my calculation is correct.

File: examples/fixed_axis_tsne.py (hunk 7 above)
Comment (qiaodan97) [reply examples/fixed_axis_tsne.py:247]: I am copy & past it from our force scheme file. I am doing this because this is just an example and I tried to make it able to run individually and without any dependency. We can definitely change it to depend on the older file, if that's preferred. Also, the original tsne code is grabbed from internet and the source code is the github link I sent to you earlier in teams. I added the citation of the original code in the file.

File: examples/fixed_axis_tsne.py (hunk 8 above)
Comment (qiaodan97) [reply examples/fixed_axis_tsne.py:235]: Same as above.

File: neo_force_scheme/neo_force_scheme.py (hunk 3 above)
Comment (christinoleo) [reply neo_force_scheme/neo_force_scheme.py]: got it! I think it is better to change its name then.. z-score in machine learning is something similar to the kruskel score (another way of giving a final score to a model or a visualization). This will confuse a bunch of people. Can you change the variable name?

File: neo_force_scheme/neo_force_scheme.py (hunk 4 above)
Comment (christinoleo) [reply neo_force_scheme/neo_force_scheme.py:173]: I think then it is good to include what the value of condifence_interval means/is used to in the :params comment. Can you do that?

File: examples/fixed_axis_tsne.py (hunk 9 above)
Comment (qiaodan97) [reply examples/fixed_axis_tsne.py:300]: Same as above. this is just meant to be an example of fixed axis tsne, we can definitely combine it into the fixed axis force scheme package if that's preferred.

File: neo_force_scheme/neo_force_scheme.py (hunk 5 above)
Comment (christinoleo) [reply neo_force_scheme/neo_force_scheme.py]: Also, I'd add in the :params comment what are good values for condifence_interval. Such as 1.0 if (bla) and 0.9 if (something else). Can you add that too please?

File: examples/example_plotly_fixed_axis_moving_in_range.py (hunk 6 above)
Comment (christinoleo) [reply examples/example_plotly_fixed_axis_moving_in_range.py:19]: This is a nice explanation. In the previous comment, I said something to include in the :param comment. Maybe this could be it :)

File: examples/fixed_axis_tsne.py (hunk 7 above)
Comment (christinoleo) [reply examples/fixed_axis_tsne.py:247]: Yeah, it is better to use things already implemented. If we ever need to move or update it, then we would need to find all the different implementations and update them individually.

File: neo_force_scheme/neo_force_scheme.py (hunk 3 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py]: variable name is changed to gaussian_function_z_score

File: neo_force_scheme/neo_force_scheme.py (hunk 4 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py:173]: Comment added at neo_force_scheme.py line 330 - 340

File: neo_force_scheme/neo_force_scheme.py (hunk 5 above)
Comment (qiaodan97) [reply neo_force_scheme/neo_force_scheme.py]: Comment added at neo_force_scheme.py line 330 - 340

File: examples/fixed_axis_tsne.py (hunk 7 above)
Comment (qiaodan97) [reply examples/fixed_axis_tsne.py:247]: license and reference added

**Review by christinoleo — APPROVED** [review]
